### DISCUSSION TITLE
How to save record.refreshTokenKey()?

### DISCUSSION THREAD

**Discussion by AfdhalAtiffTan** [pocketbase/pocketbase#4268] [Q&A] [answered]
Hello,

I am trying to sync admins' and users' tokens between multiple Pocketbase instances using the built-in router:

```js

routerAdd("PATCH", "/refreshTokenKey", (c) => {
    const data = $apis.requestInfo(c).data;

    const dao = $app.dao().withoutHooks();

    // data.table = "users"
    const record = dao.findAuthRecordByUsername(data.table, data.username);

    /* manual test */
    // record.setTokenKey(data.tokenKey);

    /* auto test */
    console.log(record.tokenKey()); // old token
    record.refreshTokenKey();
    console.log(record.tokenKey()); // new token

    dao().saveRecord(record); // store new token

    return c.json(200, data);
});
```

Calling saveRecord() doesn't appear to work (returned 400 instead of 200 code).

Any advice?

**Comment by ganigeorgiev** [comment]
I don't understand what do you mean by _"sync admins' and users' tokens between multiple Pocketbase instances"_.

PocketBase is fully stateless and doesn't store the tokens in the database. The `tokenKey` available with each user record is a partial key that takes part of the user JWT secret generation.

> Calling saveRecord() doesn't appear to work (returned 400 instead of 200 code).

You can find out why it fails by checking the logs in the _Admin UI > Logs_ and/or start the executable with `--dev`.

**Reply by AfdhalAtiffTan** [reply]
Hello Gani,

Sorry, I meant to say tokenKey, not token.

This is related to my last question: [#4241](https://github.com/pocketbase/pocketbase/discussions/4241)

I have looked at Marmot SQLite replicator, I liked it but I can't select which table to replicate.

Since I can't access any attached external database via the API (Pocketbase only support one DB), the following is just a workaround:
I have two Pocketbase instances running from `/var` (on-disk) and from `/tmp` (in-memory).

With the same id, username and password, I am trying to use the same api-token on both instances.

The log is:
```json
{
  "id": "4n1twa4krov1mww",
  "created": "2024-01-31 09:31:29.875Z",
  "updated": "2024-01-31 09:31:29.875Z",
  "data": {
    "auth": "guest",
    "details": "TypeError: Not a function: [object Object] at <eval>:18:8(39)",
    "error": "Something went wrong while processing your request.",
    "execTime": 0.523,
    "method": "PATCH",
    "referer": "",
    "remoteIp": "127.0.0.1",
    "status": 400,
    "type": "request",
    "url": "/syncSalt",
    "userAgent": "Go-http-client/1.1",
    "userIp": "127.0.0.1"
  },
  "message": "PATCH /syncKey",
  "level": 8
}
```

Does this mean a record returned by findAuthRecordByUsername() has no saveRecord()?

Now that you mention it being a partial key, I am not sure if I can proceed with the workaround.

If this is not possible, I guess my next option would be to use Sqlite-CLI to modify the tokenKey record manually.
I really want to avoid this as I think I will need to stop one of the instances momentarily just to sync the tokenKey.

**Reply by ganigeorgiev** [reply, marked as answer]
On second look I think the error in your code is the `dao().saveRecord(record)` call - it must `dao.saveRecord(record)`.

**Reply by ganigeorgiev** [reply]
Also note that you'll need to sync the app settings across the instances as the second part of the secret is stored there (the settings are stored as serialized json in the `_params` table).

**Reply by AfdhalAtiffTan** [reply]
Thanks Gani for the speedy reply and the mention of secrets in `_params` table. I find this to be very helpful.

You are right, it should be `dao.saveRecord(record)`, silly me😅
Thank you!

I think I am very close to being able to sync `tokenKey`s between multiple instances.

I have another question.

In `ext.js`:
```js
module.exports.user = (e) => {
	const record = $app.dao().findRecordById(e.model.tableName(), e.model.id);

	const res = $http.send({
		url: "http://localhost:8080/syncTokenKey",
		method: "PATCH",
		headers: { "content-type": "application/json" },
		body: JSON.stringify({
			table: e.model.tableName(),
			id: e.model.id,
			tokenKey: record.get("tokenKey")
		})
	});

	console.log(res.statusCode);
};
```

In `main.pb.js`:
```js
onModelAfterUpdate((e) => {
    const ext = require(`${__hooks}/ext.js`);
    console.log("onModelAfterUpdate: users, _admins");
    ext.user(e);
}, "users", "_admins");

onModelAfterCreate((e) => {
    const evt = require(`${__hooks}/ext.js`);
    console.log("onModelAfterCreate: users, _admins");
    ext.user(e);
}, "users", "_admins");

routerAdd("PATCH", "/syncTokenKey", (ctx) => {
    const dao = $app.dao().withoutHooks();
    const data = $apis.requestInfo(ctx).data;
    const record = dao.findRecordById(data.table, data.id);

    record.set("tokenKey", data.tokenKey);
    dao.saveRecord(record);

    return ctx.noContent(204);
});
```

It is working for `users` table, but not for `_admins` table. 

The log:
```
GoError: sql: no rows in result set at reflect.methodValueCall (native)
```

It seems like I can read `users` table but not `_admins` table.

Any advice?

**Reply by ganigeorgiev** [reply]
`Admin` and `Record` are 2 separate entities, aka. `findRecordById` works only for collection records. For admins you can use `findAdminById(id)` (also the `Admin` model is a typed Go struct so there are no `get("tokenKey")` method there but a `tokenKey` field).

_Note that this most likely will change in the future as there are plans to refactor the models and make the base model `Record`._

**Reply by AfdhalAtiffTan** [reply]
Thank you Gani!
